### ISSUE TITLE
Rewrite in TypeScript

### ISSUE THREAD

**Issue by SimenB** [jest-community/eslint-plugin-jest#256]
Now that https://github.com/typescript-eslint/typescript-eslint/pull/425 has been released, it should be relatively straightforward to migrate.

I think porting one and one rule makes the most sense. Feel free to pick one 🙂 

- [x] `consistent-test-it` - #327
- [x] `expect-expect` - #325
- [x] `lowercase-name` - #258
- [ ] `no-alias-methods`
- [x] `no-commented-out-tests` - #305
- [x] `no-disabled-tests` - #315
- [x] `no-duplicate-hooks` - #318
- [ ] `no-empty-title`
- [x] `no-export` - #323
- [x] `no-focused-tests` - #314
- [x] `no-hooks` - #322
- [ ] `no-identical-title`
- [x] `no-if` - #324
- [x] `no-jasmine-globals` - #315
- [x] `no-jest-import` - #259
- [ ] `no-large-snapshots`
- [x] `no-mocks-import` - #309
- [x] `no-test-callback` - #321
- [x] `no-test-prefixes` - #328
- [x] `no-test-return-statement` - #320
- [ ] `no-truthy-falsy`
- [ ] `prefer-called-with`
- [ ] `prefer-expect-assertions`
- [x] `prefer-inline-snapshots` - #319
- [x] `prefer-spy-on` - #326
- [x] `prefer-strict-equal` - #329
- [ ] `prefer-to-be-null`
- [ ] `prefer-to-be-undefined`
- [ ] `prefer-to-contain`
- [ ] `prefer-to-have-length`
- [x] `prefer-todo` - #335
- [ ] `require-tothrow-message`
- [x] `valid-describe` - #308
- [ ] `valid-expect-in-promise`
- [ ] `valid-expect` - #333

**Comment by G-Rath** [comment]
I've done a first pass of (`no-mocks-import`](https://github.com/G-Rath/eslint-plugin-jest/tree/ts-migration/no-mocks-import), which passes all the tests.

So far, it seems to me like this'll be a bit teeth pully, as `ts-estree` doesn't provide you with lots of nice [type guards like TS does](https://github.com/microsoft/TypeScript/blob/master/src/compiler/utilities.ts#L5392) - as a result, I've started to make my own in `tsUtils`, but am trying to be conservative for now, since I don't know `ts-estree` well enough to predict how wild things might get.

I've also encapsulated the actual primary condition into a `isMockImportLiteral` function - let me know if you're happy with how I've done that; I've never worked in the FB/jest domain, so I don't know what you guys prefer for coding in general :joy:

Finally, off the bat there's already room for improvement, which ties in to everything I've said so far: right now the rule is only checking actual string literals, which means template strings already won't work right off the bat (regardless of if they've got actual embedded expressions or not)

This could be improved to support them w/o expressions, but you could also take this further to try and provide basic support for expressions, since you might be able to do a basic search of surrounding nodes in an attempt to see if the expression is a constant, and if it is evaluate its value.

All in all, for now I've opted for doing the basics, w/ the idea that people can fill issues requesting improvements, just to keep things simple.

I've not made a PR since I saw your previous PRs were reverted following some breaking of peoples builds. Let me know if you're still wanting to go ahead with converting to TS.

For now I'm going to continue trying to pick off the easy rules - let me know any thoughts you have :)

**Comment by jeysal** [comment]
@G-Rath thanks for investing time in this! What you've described sounds fine to me. @SimenB will probably want to kick off a second attempt at TS migration at a less busy time

**Comment by SimenB** [comment]
Thanks @G-Rath! Yeah, it seems using the module from https://github.com/typescript-eslint/typescript-eslint/pull/425 requires consumers to use typescript as well, so I need to figure out how we can _author_ rules in TS without consumers noticing at all. I'll try to get to this, but I'm still swamped at work... Feel free to dig into it if you don't mind!

**Comment by G-Rath** [comment]
@SimenB while things like AST have always fascinated me, sadly I've not had a lot of time (yet - hopefully 😉) to actually use such technology in a meaningful way - up til now it's been "wide eye boy pressed up to the glass of the store window" for me :joy:

I mention that just to let you know where I sit on the skill-base for doing this stuff - so take what I say with a grain of salt, b/c it's all still very new from a usage POV.

From what you've said, and from what I've done converting `no-mocks-import`, I think we're doing this the wrong way?

My thinking is if you're using the `@typescript-eslint` packages in your (as in `eslint-plugin-jest`), then that means you're using their parser - which is a mistake, since that parser is meant for TypeScript. Hence your problem :)

What I think you're wanting is to instead just convert the existing rules into TS, making them typesafe, without changing the underlying implementation that handles the parsing (which is like `eslint`?) - for me the biggest red flag is that I was changing the methods actually used to do analysis - I was looking at the AST in ASTViewer for `@typescript-eslint/parser`, which means thats whats doing the parsing, meaning its expecting TypeScript...

I'm going to stop typing here b/c I think I've written enough for you to hopefully understand what I'm getting at.

Minor disclaimer: My mind's still a bit fuzzy, as I had wisdom teeth surgery this morning, but I really wanted to get this written down - sorry for the rambling; I'll do a second pass on this comment in a couple of days to clean it up :joy:

----

Bottom line: Pretty sure I know why it's breaking for everyone, and how to fix it. I'll give it a crack if I can get the time :)

(It's also possible that I'm completely wrong, either b/c I don't really know how `eslint` works on an implementation level, or just that I've missed what you're actually to do - feel free to push back or shush me 😄 )

**Comment by SimenB** [comment]
>I don't know enough about the eslint ecosystem to know if its worth using the `@typescript-eslint/parser` over the current one, which is now my question - what is the advantage?

We don't want to use the parser in our own rules - we want to use the utils which provide way better types than the ones from DefinitelyTyped as well as some nice utils for options etc. It's also stricter (also called opinionated 😛), forcing us to follow some good conventions (fixing both #201 and #202 at once).

We do however want to use the parser for linting our own code, but that should not be visible to consumers.

---

The dependency on `typescript` was not on purpose, see https://github.com/typescript-eslint/typescript-eslint/pull/425#issuecomment-498302492

When a release is out, we can probably retry this (I'll revert my earlier revert). Will need to dig into #268 as well at that point.

**Comment by M4rk9696** [comment]
@SimenB I'll pick up a rule, can you tell me where code migrated to typescript is maintained.

**Comment by SimenB** [comment]
A new try is up here: #305.

@M4rk9696 thanks! A PR to that branch is most welcome

**Comment by SimenB** [comment]
@G-Rath would you be able to PR `no-mocks-import` to the `reapply-ts` branch of this repo? 🙂

**Comment by G-Rath** [comment]
@SimenB I can but try ;P

Give me a few minutes (or an hour, depending on the work - I've not actually been following the activity 😬)

**Comment by SimenB** [comment]
There is absolutely no rush 🙂

**Comment by G-Rath** [comment]
@SimenB done in #309 

It took a bit longer than a few minutes b/c of a standup meeting & having to update the code a bit (but nothing major).

All the tests are passing, so I think it's fine :)

**Comment by G-Rath** [comment]
I've been doing everything in a [single generic branch](https://github.com/G-Rath/eslint-plugin-jest/tree/ts-migration/general), and have about a quarter of the rules converted?

My plan is to just cherry-pick at the end. There are a couple of that don't require `tsUtils` that I can cherry-pick, which I'll try and do now.

Otherwise, I'll finish off the rest off the rules this evening/weekend :)

**Comment by SimenB** [comment]
Oh wow @G-Rath, that's awesome!

**Comment by G-Rath** [comment]
@SimenB just a heads up that the `meta` property should be checked on all my PRs; I've been forgoing trying to figure out what the best values for the description, setting the correct "recommended" value, the "category", and the "type" in favor of converting the actual rules as I feel those are somewhat arbitrary.

I've been just copying the header for each rule as the `description` :)

I've also not tested any of the fixes - I think they'll work fine but the only thing I'm worry about is what happens if you return an empty array..? I've not written eslint plugins before so don't know what the best practices are.

I'm happy to follow your lead on any changes you want made.

**Comment by SimenB** [comment]
Ah, ok, good to know. I'll verify

The fixes are all covered by unit tests, so if the tests pass then we're good. If a violation is not fixable, we should not return a fixer (that way the "n error potentially fixable" message from eslint is not misleading)

**Comment by G-Rath** [comment]
I think we've nearly gotten all the low hanging fruit, so I'm going to look to tackle the converting of the `exportCase` & other type guard util functions (like `method`, `method2`, etc).

I've actually already done `exportCase`, but I've not pushed it yet b/c I need to explore them a bit more, as some of them are definitely guards, while others are not.

Once that's done, we'll be able to tackle all of the `prefer-*`, as they're the main rules that consume those methods.

I've got a few more rules to convert & cherry pick first tho - I'm catching a flight in a day or so, so I might go dark for a bit, but at this rate I'd say we should easily be done by the end of month, if not by next week :D

**Comment by SimenB** [comment]
Haha, this has been awesome, thank you so much for your help!

**Comment by SimenB** [comment]
TS ESLint had a stable release, so I merged the PR to master

**Comment by G-Rath** [comment]
> I merged the PR to master

Well, this went about as well as I expected - I woke up to about 20 emails for this repo, half of which being semantic bot :joy:

but seriously it actually did go as expected: the bugs I've seen are pretty much what I expected, and the fixing PRs have been great, and will make everything more stable.

I'll look to power through getting the rest of the rules converted over the next couple of days.
Pretty much all the rules that are left to do I've got half converted, as I'm picking the ones w/ the least amount of imports from `tsUtils` first.

**Comment by SimenB** [comment]
Heh, yeah... It wasn't too bad, though!

**Comment by G-Rath** [comment]
@SimenB Just wanted to let you know I've not forgotten about this 😉 

I've got 2 rules left to convert, and the new guards + expect parser is working a treat.

Life got a bit busy once I got back, but I hope to have a PR for review by midweek.

As an example of how nice the new expect stuff is: `prefer-to-be-null` & `prefer-to-be-undefined` took literally 5 minutes to convert, despite being the two rules that used those massive expect guards, as they were having to account for `not`.

Now they're just:

```
const isNullEqualityMatcher = (
  matcher: ParsedExpectMatcher,
): matcher is ParsedNullEqualityMatcher =>
  EqualityMatcher.hasOwnProperty(matcher.name) &&
  matcher.node.parent !== undefined &&
  matcher.node.parent.type === AST_NODE_TYPES.CallExpression &&
  hasOneArgument(matcher.node.parent) &&
  isNullLiteral(matcher.node.parent.arguments[0]);

/* in create: */
if (!isExpectCall(node)) {
  return;
}

const { matcher } = parseExpectCall(node);

if (matcher && isNullEqualityMatcher(matcher)) {
  context.report({
    fix(fixer) {
      return [
        fixer.replaceText(matcher.node.property, 'toBeNull'),
        fixer.remove(matcher.node.parent.arguments[0]),
      ];
    },
    messageId: 'useToBeNull',
    node: matcher.node.property,
  });
}
```

**Comment by SimenB** [comment]
This is super exciting 😀

**Comment by G-Rath** [comment]
Except I missed the fact that you can chain modifiers (i.e `resolve.not`) 😭

Not a big deal since we only really check for them in `valid-expect`, but means I have to code up a linked-list style loop to the parser which are always a bit tedious in TS

**Comment by SimenB** [comment]
You can only have `expect().not.toBe()`, `expect().resolves.toBe()`, `expect().rejects.toBe()`, `expect().resolves.not.toBe()` and `expect().rejects.not.toBe()`. So I don't think we necessarily need a linked list? If we can say a parsed `expect` may or may not be `not`, and it might be either `resolves` or `rejects`. Dunno, might make it hard to do `.parent` etc?

**Comment by G-Rath** [comment]
Interesting. The typescript types allow it, but it fails at runtime: http://puu.sh/DYIn4/7278eb6827.png

When I say "linked list" it's not anything complex, just:

```
  let node = expectCall.parent;
  let previousModifier: { modifier?: ParsedExpectModifier } = expectation;

  while (node.parent && isExpectMember(node.parent)) {
    const member = parseExpectMember(node.parent);

    if (isParsedExpectModifier(member)) {
      previousModifier = previousModifier.modifier = member;
    }

    node = node.parent;
  }
```

It was more a bit of a shock b/c I'd not considered it, and so it wasn't in the output 😬 
Doing it this way means I don't have to refactor anything else, and also has the bonus of requiring the previous modifier to be defined (otherwise `parsedExpect.modifer !== undefined && parsedExpect.modifier2 !== undefined`)

**Comment by SimenB** [comment]
That should probably be a type error, yeah. The Jest team doesn't maintain those typings, but Jest's internal typings are probably wrong in this regard as well. I'll make a point to fix it before releasing Jest 25

**Comment by G-Rath** [comment]
Ahh I see how it is - I had a proper read over everything, and yeah that loop while nice is overkill (plus the types are murder):

```
  describe('each', () => {
    it.each([
      expect(true).not.resolves,  // true
      expect(true).not.rejects,    // true
      expect(true).resolves.not, // false
      expect(true).rejects.not    // false
    ])('should be undefined', (result) => {
      expect(result).toBe(undefined);
    });
  });
```

So the allowed modifiers are:

`not`
`resolves`
`rejects`
`resolves.not`
`rejects.not`

However, `valid-expect` doesn't actually check for that. In fact, it has those in it's tests as *valid*.
In fact, the majority of it's `not` valid tests have `not` *first*.

I'll avoid "fixing" that if possible currently, and instead look into it after I've gotten the remaining rules converted & merged.

But in light of this I think the typings can be a lot simpler. I have a few ideas on a couple of different ways it could be typed, and will have a play around to find out what works best :)

**Comment by G-Rath** [comment]
```
------------------------------|----------|----------|----------|----------|-------------------|
File                          |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
------------------------------|----------|----------|----------|----------|-------------------|
All files                     |    96.26 |    92.62 |    89.54 |     96.4 |                   |
 src                          |      100 |      100 |      100 |      100 |                   |
  index.js                    |      100 |      100 |      100 |      100 |                   |
 src/processors               |      100 |      100 |      100 |      100 |                   |
  snapshot-processor.js       |      100 |      100 |      100 |      100 |                   |
 src/rules                    |    96.21 |    92.55 |    89.26 |    96.35 |                   |
  consistent-test-it.ts       |      100 |      100 |      100 |      100 |                   |
  expect-expect.ts            |      100 |      100 |      100 |      100 |                   |
  expectGuards.ts             |    72.73 |    47.27 |    53.33 |    72.73 |... 17,424,425,450 |
  final.ts                    |    92.54 |    83.87 |       84 |    92.42 |... 43,365,367,433 |
  lowercase-name.ts           |      100 |      100 |      100 |      100 |                   |
  no-alias-methods.ts         |      100 |      100 |      100 |      100 |                   |
  no-commented-out-tests.ts   |      100 |      100 |      100 |      100 |                   |
  no-disabled-tests.ts        |      100 |      100 |      100 |      100 |                   |
  no-duplicate-hooks.ts       |      100 |      100 |      100 |      100 |                   |
  no-empty-title.ts           |      100 |      100 |      100 |      100 |                   |
  no-export.ts                |      100 |      100 |      100 |      100 |                   |
  no-focused-tests.ts         |      100 |      100 |      100 |      100 |                   |
  no-hooks.ts                 |      100 |      100 |      100 |      100 |                   |
  no-identical-title.ts       |      100 |      100 |      100 |      100 |                   |
  no-if.ts                    |      100 |      100 |      100 |      100 |                   |
  no-jasmine-globals.ts       |      100 |      100 |      100 |      100 |                   |
  no-jest-import.ts           |      100 |      100 |      100 |      100 |                   |
  no-large-snapshots.js       |      100 |      100 |      100 |      100 |                   |
  no-mocks-import.ts          |      100 |      100 |      100 |      100 |                   |
  no-standalone-expect.ts     |      100 |      100 |      100 |      100 |                   |
  no-test-callback.ts         |      100 |      100 |      100 |      100 |                   |
  no-test-prefixes.ts         |      100 |      100 |      100 |      100 |                   |
  no-test-return-statement.ts |      100 |      100 |      100 |      100 |                   |
  no-truthy-falsy.ts          |      100 |      100 |      100 |      100 |                   |
  no-try-expect.ts            |      100 |      100 |      100 |      100 |                   |
  prefer-called-with.ts       |      100 |      100 |      100 |      100 |                   |
  prefer-expect-assertions.ts |      100 |      100 |      100 |      100 |                   |
  prefer-inline-snapshots.ts  |      100 |      100 |      100 |      100 |                   |
  prefer-spy-on.ts            |      100 |      100 |      100 |      100 |                   |
  prefer-strict-equal.ts      |      100 |      100 |      100 |      100 |                   |
  prefer-to-be-null.ts        |      100 |      100 |      100 |      100 |                   |
  prefer-to-be-undefined.ts   |      100 |      100 |      100 |      100 |                   |
  prefer-to-contain.ts        |      100 |      100 |      100 |      100 |                   |
  prefer-to-have-length.ts    |      100 |      100 |      100 |      100 |                   |
  prefer-todo.ts              |      100 |      100 |      100 |      100 |                   |
  require-tothrow-message.ts  |      100 |      100 |      100 |      100 |                   |
  tsUtils.ts                  |    89.86 |    75.38 |    73.08 |    89.71 |... 42,145,166,178 |
  util.js                     |       65 |        0 |    22.22 |    70.59 |     7,10,15,20,29 |
  valid-describe.ts           |      100 |      100 |      100 |      100 |                   |
  valid-expect-in-promise.ts  |      100 |      100 |      100 |      100 |                   |
  valid-expect.ts             |      100 |    98.75 |      100 |      100 |               274 |
------------------------------|----------|----------|----------|----------|-------------------|
```

Guess who got `valid-expect` typesafe last night ;)

I've just got `no-large-snapshots` to convert, which will be interesting b/c it uses `babel-eslint` - hopefully it'll have TS typings, but if not I might have to leave that one to you @SimenB 

I expect to have a PR by EOD today (~8 hours) 😄

**Comment by SimenB** [comment]
This is done! Thank you to everyone who contributed, and especially @G-Rath who did all the hard work 🎉
